tests/unit/test_ecies.cpp

```diff
@@ -0,0 +1,186 @@
+    std::string publicKey;
+    std::string ciphertext;
+    std::string plaintext;
+    std::string macTag;
+    std::string sessionKey;
+};
+
+TEST(ECIESTest, testWithBackendData)
+{
+    std::string privKeyString = \
+            "-----BEGIN EC PRIVATE KEY-----\n"
+            "MIGkAgEBBDBUGq/lzm93hmnIVdr5UWMnSnnN1Kd0EVkHxRJ0BFzQcUJxrH579/04\n"
+            "jQbIfYP4VPCgBwYFK4EEACKhZANiAARl2UGjmJow8rUCXZkGXTlRnhQXImCC7ISO\n"
+            "nqzT64Nzpfzsnv3FsNjjk5MCh4WfZRaZwApNyWLX5HwZhF8ZOgC41Xryd8pBREKu\n"
+            "Qqd2w1/Fb1aRsmmI893lRDwb0qRb5Ug=\n"
+            "-----END EC PRIVATE KEY-----\n";
+
+    struct testData data;
+    data.publicKey =  "0437D97E1299CF430163A11FC87B7B8FB647BD98EE8B5BAE31C3BB1EE897A77BAFA20F94F6CE0B91748996359000B47E"
+                      "56F2587825E4F01CEFF9BBDDE8ED17DDEE8BD46BD409DFF7FF48E9F7A869FE3F9D8174F788CD21C2D8EAE5BF256884AA"
+                      "89";
+    data.ciphertext =  "90C4509BC7C9FFE404F6B063E2781655323BF402645C241A5F57E556AB397E6C4EFED563A77A66998A4645C7A24B82"
+                       "34";
+    data.plaintext =  "4AB1D2A13A956C83D58B0AD023115256F8720C7E6E332B47906D5D0BA5257DE7";
+    data.macTag =     "46805E4419BEB16F7EC8D234699A1E266398819B46AD095AF7A49CC683E7561ABFB24E3DF5798B3E1451110A8F663F15"
+                      "47B640CE65112051F4F8481DCE578F4D";
+
+    /* Read and set the private key */
+    AsymmetricKeypair privKey(AsymmetricKeypair::readPrivateKeyFromPEM(privKeyString, ""));
+    ECIESDecryptionCtxBuilder decBuilder(privKey);
+
+    /* Set constants */
+    const SymmetricCipherMode mode = SymmetricCipherMode::CBC;
+    const SymmetricCipherKeySize keySize = SymmetricCipherKeySize::S_256;
+    const SymmetricCipherPadding padding = SymmetricCipherPadding::PKCS;
+    const std::vector<uint8_t> zeroIV = std::vector<uint8_t>(AESCipherBuilder::getDefaultIVLength(mode));
+
+    /* Create KDF context */
+    auto kdf = std::make_shared<X963KDF>(X963KDF(openssl::DigestTypes::SHA512));
+
+    /* Create the decryption context */
+    decBuilder.setKDF(kdf)
+              .setMacFactoryFunction([](const std::vector<uint8_t> &key) {
+                    return std::make_unique<mococrw::HMAC>(mococrw::HMAC(openssl::DigestTypes::SHA512, key));
+                })
+              .setMacCipherKeySize(512 / 8)
+              .setSymmetricCipherFactoryFunction([zeroIV](const auto &key, const auto &operation) {
+                    auto builder = mococrw::AESCipherBuilder(mode, keySize, key).setIV(zeroIV);
+                    builder.setPadding(padding);
+                    if (operation == SymmetricCipherOperation::Decryption) {
+                        return builder.buildDecryptor();
+                    }
+                    return builder.buildEncryptor();
+                })
+              .setSymmetricCipherKeySize(256 / 8);
+    auto decCtx = decBuilder.buildDecryptionCtx();
+
+    /* Set the received mac value */
+    decCtx->setMAC(utility::fromHex(data.macTag));
```

Comment:
Same thing here, please use setEphemeralKey, update, setMAC, finish in this order.

Reply:
Done